component/block_cache/block_cache.go

```diff
@@ -1213,12 +1213,13 @@ func (bc *BlockCache) waitAndFreeUploadedBlocks(handle *handlemap.Handle, cnt in
 		block := node.Value.(*Block)
 		if block.id != -1 {
 			// Wait for upload of this block to complete
-			<-block.state
+			_, ok := <-block.state
+			if ok {
+				block.Unblock()
+			}
 			block.flags.Clear(BlockFlagUploading)
 		}
 
-		block.Unblock()
```

Comment:
block-id can be -1 and for those we need to close the channel as well but if id is  != -1 then we shall not wait to receive anything. Hence the code was written like this. block-id == -1 is a valid case where due to prefetch we allocated more blocks then required and we never used them to fill data as there offsets were beyond the file length. So may be better fix is to release those blocks when you try to do lineupDownload() and function returns EOF. That will fix this problem too.